src/data-fetchers/bam/bam-data-fetcher.ts

```diff
@@ -76,8 +79,8 @@ class BamDataFetcher {
         (await this.worker).fetchTilesDebounced(this.uid, tileIds).then(receivedTiles);
     }
 
-    async getTabularData(uid: string, tileIds: string[]): Promise<Segment[] | SegmentWithMate[] | Junction[]> {
-        const buf = await (await this.worker).getTabularData(uid, tileIds);
+    async getTabularData(tileIds: string[]): Promise<BamTile[]> {
```

Comment:
Since `Segment[] | SegmentWithMate[] | Junction[]` is a more scoped type than `BamTile[]`, we could define a dedicated type, e.g., `BamTiles`:

```ts
// maybe in gosling.schema.guards.ts
type UnionOfArray<T> = T extends infer I ? I[] : never;

// bam-worker.ts
type BamTile = Junction | SegmentWithMate | Segment;
type BamTiles = UnionOfArray<BamTile>; // Junction[] | SegmentWithMate[] | Segment[]

// bam-data-fetcher.ts
async getTabularData(tileIds: string[]): Promise<BamTiles> {
```

We could use `BamTiles` in https://github.com/gosling-lang/gosling.js/blob/83cd2350945c2f4cefa2bdc667cb2a777795948e/src/data-fetchers/bam/bam-worker.ts#L437 as well

```suggestion
    async getTabularData(tileIds: string[]): Promise<BamTiles[]> {
```

Reply:
I see what you are saying.  I have a slightly different idea using generics:

```typescript
type InferTileType<T extends BamData> =
	T['extractJunction'] extends true ? Junction :
	T['loadMates'] extends true ? SegmentWithMate :
	Segment;

class BamDataFetcher<T extends BamData = BamData> implements TabularDataFetcher<InferTileType<T>> {
```

This way we encode the relationship between config and the expected tile type.

```typescript
declare var fetcher: BamDataFetcher;
fetcher.getTabularData([]) // Promise<Segment[] | SegmentWithMate[] | Junction[]>;

declare var junctionFetcher: BamDataFetcher<{ type: 'bam', url: string, indexUrl: string, extractJunction: true }>;
fetcher.getTabularData([]) // Promise<Junction[]>;
```